Shopify/cli
# [CLI UI kit - Themes adoption] Extract the ownership of development themes

### WHY are these changes introduced?

Fixes https://github.com/Shopify/internal-cli-foundations/issues/509
Fixes https://github.com/Shopify/cli/issues/1084
Fixes https://github.com/Shopify/cli/issues/1263

Requires https://github.com/Shopify/shopify-cli/pull/2724

### WHAT is this pull request doing?

* Generates Development Theme ID and calls Themes API in Node.js CLI.
* Depending on the `shopify theme` command: passes ID of generated theme to Ruby CLI for further processing.

### How to test your changes?

Make sure that your local Ruby CLI uses changes from https://github.com/Shopify/shopify-cli/pull/2724

1. Verify that `(pnpm run )shopify theme info` says `Development Theme ID   Not set`
2. Create a new development theme: `shopify theme dev --path="./my-theme"`
3. Verify that `shopify theme info` now displays the ID for the created dev theme
4. Verify that `shopify theme open -d` opens the dev theme (with the correct ID)
5. Verify that `shopify theme delete -d` deletes the dev theme (after you confirm the deletion)
6. Verify that `shopify theme info` says `Development Theme ID   Not set` again
7. Verify that `shopify theme pull -d --path="./my-theme"` throws the error `No development theme ID has been set. Please create a development theme first.`
8. Verify that `shopify theme push -d --path="./my-theme"` creates a new development theme.
9. Verify that `shopify theme info` now displays the ID for the created dev theme
10. Verify that `shopify theme dev --path="./my-theme"` uses the correct ID
11. Delete the theme with CLI 2.x (so that 3.x doesn't know anything about it): `shopify(-dev) theme delete <your ID>`
12. Verify that `shopify theme dev --path="./my-theme"` still works, but has created a new ID
13. Delete this theme also with CLI 2.x
14. Verify that `shopify theme pull -d --path="./my-theme"` throws the error `Development theme <old ID> could not be found. Please create a new development theme.`

#### `--theme-editor-sync`
1. `shopify theme init` a new theme.
2. Make sure that no development theme ID is set in CLI.
3. `shopify theme dev --store="…" --path="./draft" --theme-editor-sync`
4. Verify that CLI does not display any "Keep the remote version" prompt.

#### Keep Development Theme ID when switching stores
1. Execute `shopify theme dev --path="./my-theme" --store=<store 1>` and stop the server
2. Note down the theme ID that was created
3. Execute `shopify theme dev --path="./my-theme" --store=<store 2>` and stop the server
4. Note down the theme ID that was created
5. Execute `shopify theme dev --path="./my-theme" --store=<store 1>` again
6. Verify that the ID for store 1 was reused
7. Execute `shopify theme dev --path="./my-theme" --store=<store 2>` and stop the server
8. Verify that the ID for store 2 was reused

### Measuring impact

How do we know this change was effective? Please choose one:


### Checklist
